### PR TITLE
Feature/burger builder

### LINKED ISSUE #3: [Burger Order Page] Custom Burger
- [x] [ UI ] A price box at the side   
- [x] [ UI ] Click the burger and sandwitch button can switch UI to it.
- [x] [ UI ] Click the ingredient will add or decrease the number of ingredient
- [x] [ UI ] Set the min number of ingredient to 0 and max number of ingredient to 3 
- [x] [ UI ] total price is being calculated based on the number of ingredient 
- [x] [ store ] Fetch ingredient from api & initial price

### PR DESCRIPTION
**Pull request by WendyChenj**
1. define ingredient types
2. create custom burger store and initial burger state
3. fetch burger ingredient and burger price from store
4. buttons that allow customers to add or remove ingredients & burger UI based on the changes of ingredients.
5. create a price box at the side of the page and calculate the total price based on the added or removed ingredients

### CODE REVIEW

File: src/common/store/customBurgerStore/customBurgerActionTypes.ts [hunk 1]
```diff
@@ -19,4 +19,53 @@ export interface ChooseBurgerBread {
...
+  sauceName: string;
+}
+
+export interface RemoveBurgerSauce {
+  type: typeof REMOVE_BURGER_SAUCE;
+  sauceName: string;
+}
+
+export type CustomBurgerActionTypes =
+  | ChooseBurgerBread
```
Comment (hereiskeith) [inline src/common/store/customBurgerStore/customBurgerActionTypes.ts:63]: I think the first | is not necessary...

File: src/common/store/customBurgerStore/customBurgerReducer.ts [hunk 2]
```diff
@@ -35,15 +75,149 @@ const initialState: BurgerState = {
       choose: false,
     },
   ],
+  totalPrice: 2.0,
 };
 
 const customBurgerReducer = (state: BurgerState = initialState, action: CustomBurgerActionTypes): BurgerState => {
+  let ingPrice = 0;
```
Comment (hereiskeith) [inline src/common/store/customBurgerStore/customBurgerReducer.ts]: I would suggest you moving this variable to the specific cases

File: src/common/store/customBurgerStore/customBurgerReducer.ts [hunk 3]
```diff
@@ -1,28 +1,68 @@
...
-    lettuce: 0,
-  },
-  meat: {
-    beef: 0,
-    chicken: 0,
-    ham: 0,
-    bacon: 0,
-  },
+  vegetable: [
+    {
```
Comment (hereiskeith) [inline src/common/store/customBurgerStore/customBurgerReducer.ts:18]: `vegetable { tomato: { amount: 0, price: 0.5 }, lettuce: { amount: 0, price: 0.5 } }`

File: src/common/store/customBurgerStore/customBurgerReducer.ts [hunk 4]
```diff
@@ -35,15 +75,149 @@ const initialState: BurgerState = {
...
+  let ingPrice = 0;
   switch (action.type) {
     case CHOOSE_BURGER_BREAD:
       return {
         ...state,
         bread: action.bread,
       };
+    case ADD_BURGER_MEAT:
+      const addedMeat = state.meat.map((ele: Meat) => {
+        if (ele.name === action.meatName) {
```
Comment (hereiskeith) [inline src/common/store/customBurgerStore/customBurgerReducer.ts:92]: Should be refactored accordingly

File: src/pages/BurgerOrder/components/SelfMadeBurger/SelfMadeBurger.tsx [hunk 5]
```diff
@@ -1,61 +1,100 @@
...
-    if (bread !== breadName) {
-      dispatch(chooseBurgerBread(breadName));
+  const handlePrevButton = (event: React.MouseEvent): void => {
+    event.preventDefault();
+    if (buildState > 1) {
+      setBuildState(buildState - 1);
     }
   };
 
+  const burgerMeat = meat.map((ele: customBurgerTypes.Meat) => {
```
Comment (hereiskeith) [inline src/pages/BurgerOrder/components/SelfMadeBurger/SelfMadeBurger.tsx:37]: if meat state is constructed with object, the logic should be something like
```suggestion
  const burgerMeat = Object.keys(state.meat).map(ele => {
      return new Array(ele.amount).map((, index) => {
          return <Ingredient ingType={ele} key={ele + index} />
       })
    })
```

File: src/pages/BurgerOrder/components/BurgerBuilder/BurgerBuilder.tsx [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+    vegetable: state.customBurger.vegetable,
+    cheese: state.customBurger.cheese,
+    sauce: state.customBurger.sauce,
+  }));
+
+  let burgerBuilder: JSX.Element = null;
+
+  switch (buildState) {
+    case 1:
+      burgerBuilder = <BreadBuildContainer bread={bread} />;
```
Comment (hereiskeith) [inline src/pages/BurgerOrder/components/BurgerBuilder/BurgerBuilder.tsx]: I would recommend changing BreadBuildContainer -> BreadBuildStep. So as IngredientsBuildContainer, SauceBuildContainer

File: src/pages/BurgerOrder/components/BurgerBuilder/components/SauceBuildContainer/SauceBuildContainer.tsx [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+
+type SauceTypes = {
+  sauce: Sauce[];
+};
+
+const SauceBuildContainer = ({ sauce }: SauceTypes): JSX.Element => {
+  const dispatch = useDispatch();
+
+  const handleAddSauce = (e: React.MouseEvent, sauceName: string) => {
+    e.preventDefault();
```
Comment (hereiskeith) [inline src/pages/BurgerOrder/components/BurgerBuilder/components/SauceBuildContainer/SauceBuildContainer.tsx]: Material UI Button component should have already handled the default mouse event nicely

File: src/pages/BurgerOrder/components/BurgerBuilder/components/IngredientsBuildContainer/IngredientsBuildContainer.tsx [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+                  : 'button-nonactive font-weight-bold p-a-zero p-t-xxxsmall p-b-xxxsmall m-r-xxxsmall'
+              }
+              onClick={(e: React.MouseEvent) => handleAddButton(e, buildState, ele.name, ele.amount)}
+              disabled={ele.amount === 3 ? true : false}
+            >
+              +
+            </Button>
+            <span>{ele.amount}</span>
+            <Button
+              className={
```
Comment (hereiskeith) [inline src/pages/BurgerOrder/components/BurgerBuilder/components/IngredientsBuildContainer/IngredientsBuildContainer.tsx]:
```suggestion
              className={ (ele.amount === 0 ? 'button-disabled' : 'button-nonactive') + '  font-weight-bold p-a-zero p-t-xxxsmall p-b-xxxsmall m-l-xxxsmall' }
```

File: src/scss/_global.scss [hunk 9]
```diff
@@ -50,10 +50,60 @@ $directions: (top, right, bottom, left);
   }
 }
 
+// font size
+@each $name, $size in $space-sizes {
```
Comment (hereiskeith) [inline src/scss/_global.scss]: You should define a new variable like $space-sizes for this scss class generator

**Review by hereiskeith — CHANGES_REQUESTED** [review]
Overall looks good! There are a few changes you should make to optimize your code a bit

File: src/common/store/customBurgerStore/customBurgerActionTypes.ts (hunk 1 above)
Comment (WendyChenj) [reply src/common/store/customBurgerStore/customBurgerActionTypes.ts:63]: Hi, prettier does that for me.

File: src/common/store/customBurgerStore/customBurgerReducer.ts (hunk 2 above)
Comment (WendyChenj) [reply src/common/store/customBurgerStore/customBurgerReducer.ts]: I will move it out of the reducer function and treat it as a global variable.

**Review by hereiskeith — APPROVED** [review]

**Pull request merged**